File: sdks/python/aion-worker/aion_worker/context.py
"""ActivityContext for explicit heartbeat and cooperative cancellation."""

from __future__ import annotations

import asyncio
from typing import Any

from .proto import common_pb2
from .session import ActivityId, Payload, WorkerSession, WorkflowId

JSON_CONTENT_TYPE = "application/json"
WIRE_DEFAULT_ATTEMPT = 1


class ActivityContext:
    """Context passed to each activity handler."""

    def __init__(
        self,
        *,
        workflow_id: WorkflowId,
        activity_id: ActivityId,
        attempt: int = WIRE_DEFAULT_ATTEMPT,
        session: WorkerSession,
        content_type: str = JSON_CONTENT_TYPE,
    ) -> None:
        self.workflow_id = workflow_id
        self.activity_id = activity_id
        self.attempt = attempt
        self._session = session
        self._content_type = content_type
        self._cancelled = asyncio.Event()

    async def heartbeat(self, detail: Payload | Any | None = None) -> None:
        """Send explicit heartbeat progress for this activity."""

        progress = encode_detail(detail, self._content_type)
        await self._session.send_heartbeat(self.workflow_id, self.activity_id, progress)

    def is_cancelled(self) -> bool:
        """Return whether the engine has requested cooperative cancellation."""

        return self._cancelled.is_set()

    async def cancelled(self) -> None:
        """Wait until the engine requests cooperative cancellation."""

        await self._cancelled.wait()

    def _cancel(self) -> None:
        self._cancelled.set()


class ActivityCancellationHandle:
    """Internal handle used by the receive loop to deliver cancellation."""

    def __init__(self, context: ActivityContext) -> None:
        self._context = context

    def cancel(self) -> None:
        """Mark the associated context cancelled without cancelling its task."""

        self._context._cancel()


def encode_detail(detail: Payload | Any | None, content_type: str = JSON_CONTENT_TYPE) -> Payload | None:
    """Encode heartbeat/error detail without changing existing Payload content types."""

    if detail is None:
        return None
    if isinstance(detail, common_pb2.Payload):
        return detail
    from .activity import encode_json_payload

    return encode_json_payload(detail, content_type)

File: sdks/typescript/aion-worker/src/context.ts
import type { ActivityIdKey, Payload, WorkflowId } from "./session.js";

export type HeartbeatSender = (
	workflowId: WorkflowId,
	activityId: ActivityIdKey,
	progress?: Payload,
) => Promise<void>;

export interface ActivityContextOptions {
	readonly workflowId: WorkflowId;
	readonly activityId: ActivityIdKey;
	readonly attempt: number;
	readonly heartbeatSender: HeartbeatSender;
	readonly contentType?: string;
}

export class ActivityCancellationHandle {
	private cancelledValue = false;
	private readonly waiters = new Set<() => void>();

	public isCancelled(): boolean {
		return this.cancelledValue;
	}

	public cancelled(): Promise<void> {
		if (this.cancelledValue) {
			return Promise.resolve();
		}
		return new Promise<void>((resolve) => {
			this.waiters.add(resolve);
		});
	}

	public cancel(): void {
		if (this.cancelledValue) {
			return;
		}
		this.cancelledValue = true;
		const waiters = [...this.waiters];
		this.waiters.clear();
		for (const resolve of waiters) {
			resolve();
		}
	}
}

export class ActivityContext {
	public readonly activityId: ActivityIdKey;
	public readonly attempt: number;
	private readonly workflowId: WorkflowId;
	private readonly heartbeatSender: HeartbeatSender;
	private readonly cancellation: ActivityCancellationHandle;
	private readonly contentType: string;

	public constructor(
		options: ActivityContextOptions,
		cancellation: ActivityCancellationHandle = new ActivityCancellationHandle(),
	) {
		this.workflowId = options.workflowId;
		this.activityId = options.activityId;
		this.attempt = options.attempt;
		this.heartbeatSender = options.heartbeatSender;
		this.cancellation = cancellation;
		this.contentType = options.contentType ?? "application/json";
	}

	public async heartbeat(detail?: Payload | unknown): Promise<void> {
		await this.heartbeatSender(
			this.workflowId,
			this.activityId,
			detail === undefined ? undefined : payloadFromDetail(detail, this.contentType),
		);
	}

	public isCancelled(): boolean {
		return this.cancellation.isCancelled();
	}

	public cancelled(): Promise<void> {
		return this.cancellation.cancelled();
	}
}

function payloadFromDetail(detail: Payload | unknown, contentType: string): Payload {
	if (isPayload(detail)) {
		return detail;
	}
	return {
		contentType,
		bytes: new TextEncoder().encode(JSON.stringify(detail)),
	};
}

function isPayload(value: unknown): value is Payload {
	if (typeof value !== "object" || value === null) {
		return false;
	}
	const candidate = value as { contentType?: unknown; bytes?: unknown };
	return (
		typeof candidate.contentType === "string" &&
		candidate.bytes instanceof Uint8Array
	);
}
